File: python/update_given_S.py
import numpy as np
import update_given_x as update

def update_given_S(mu, sigma, S, ind_ranking):
    """
    UPDATE_GIVEN_S Updates the mean and covariance given the word ranking.

    Args:
        mu (np.array): Prior mean vector.
        sigma (np.array): Prior covariance matrix.
        S (np.array): Embeddings of words in set (matrix).
        ind_ranking (list): Index of ranking of words given by human.

    Returns:
        mu (np.array): Updated mean vector.
        sigma (np.array): Updated covariance matrix.
    """

    # Convert ind_ranking to NumPy array (ensure integer type)
    ind_ranking = np.array(ind_ranking, dtype=int)

    while len(ind_ranking) >= 2:
        #print("Before update:")
        #print("mu:", mu)
        #print("sigma:\n", sigma)
        #print("S:\n", S)
        print("ind_ranking:", ind_ranking)

        # Call update_given_x function (ensure indexing matches MATLAB)
        mu, sigma = update.update_given_x(np.copy(mu), np.copy(sigma), np.copy(S), int(ind_ranking[0]))

        # print("\nAfter update_given_x:")
        # print("mu:", mu)
        # print("sigma:\n", sigma)

        # Remove column correctly from S
        original_index = ind_ranking[0]  # Store original index
        S = np.delete(S, original_index, axis=1)

        # Update ind_ranking *before* modifying it
        ind_ranking = np.delete(ind_ranking, 0)  # Remove first element

        # Adjust indices (decrement indices greater than removed element)
        ind_ranking = np.array([i - 1 if i > original_index else i for i in ind_ranking], dtype=int)

        print("\nAfter updating S and ind_ranking:")
        print("S:\n", S)
        print("ind_ranking:", ind_ranking)
        print("-" * 50)

    # Force symmetry if necessary
    if not np.allclose(sigma, sigma.T, atol=1e-8):  # Use tight tolerance
        print("Forcing symmetry")
        sigma = (sigma + sigma.T) / 2

    return mu, sigma

File: matlab/update_given_S.m
function [mu, sigma] = update_given_S(mu,sigma, S, ind_ranking)
% UPDATE_GIVEN_S Updates the mean and covariance given the word ranking.

% We assume (Plackett-Luce model) that the probability of a ranking outcome
% is the product of the probabilities of choosing an item among the set of
% remaining items at each step, and removing top word one at a time.
%
% Input arguments:
%   mu     - Prior mean vector.
%   sigma  - Prior covariance matrix.
%   S      - Embeddings of words in set.
%   ind_ranking  - Index of raking of words given by human
%
% Returns:
%   mu     - Updated mean vector.
%   sigma  - Updated covariance matrix.

    while length(ind_ranking)>=2
        % Calling the function with the current index
        [mu, sigma] = update_given_x(mu, sigma, S,  ind_ranking(1));
        
        % Deleting the ind_ranking(1)-th column from S
        S(:, ind_ranking(1)) = [];
        
        % Updating the ind_ranking vector
        % Since we remove a column from S, we need to decrement the indices that are
        % greater than the one we just removed.
        ind_ranking = ind_ranking - (ind_ranking > ind_ranking(1));
        
        % Now we also need to remove the first element since we just processed it
        ind_ranking(1) = [];
    end
    if (~issymmetric(sigma))
        disp("Forcing symmetry")
        sigma = (sigma + sigma.')/2;
    end
end